# <center> Rounding Numbers
<br><br><br><br>
### <center> Don't Scroll Down Yet...

---
**Question**: How would you round -2.45 to the first decimal place?

(Everyone type in their number in the "Chat box"...when I say "Go" hit the Return Key)
<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

<br><br><b>
<br><br><b>
<br><br><b>

---

## Conceptual approaches for rounding numbers

### Bias methods

**Round Half Up** (i.e. bias towards more positive)
1.  7.6 $\rightarrow$ 8
2. 7.5 $\rightarrow$ 8
3. 7.4 $\rightarrow$ 7
<br><br>
4. -7.6 $\rightarrow$ -8
5. -7.5 $\rightarrow$ -7
6. -7.4 $\rightarrow$ -7

**Round Half Down** (i.e. bias towards more negative)
1.  7.5 $\rightarrow$  7
2. -7.5 $\rightarrow$ -8

**Round Half To Zero** (i.e. bias towards zero)
1.  7.5 $\rightarrow$  7
2. -7.5 $\rightarrow$ -7

**Round Half Away from Zero** (i.e. bias towards $\pm \infty$)
1. 7.5 $\rightarrow$  8
2. -7.5 $\rightarrow$ -8

Additional information:

https://en.wikipedia.org/wiki/Rounding\#Types\_of\_rounding


### Free from positive/negative bias and bias toward/away from zero

**Round Half To Even** (i.e. zero is considered evenly distributed)
1.  7.5 $\rightarrow$  8
2.  8.5 $\rightarrow$  8
<br><br>
3. -7.5 $\rightarrow$ -8
4. -8.5 $\rightarrow$ -8


**Round Half To Odd** (i.e. zero is considered evenly distributed)
1.  7.5 $\rightarrow$  7
2.  8.5 $\rightarrow$  9
<br><br>
3. -7.5 $\rightarrow$ -7
4. -8.5 $\rightarrow$ -9

---

## Different approaches within Python

---
### Python's built-in function 'round' (with examples)

https://docs.python.org/3/library/functions.html#round

In [1]:
print( round(0.5) )
print( round(1.5) )
print( round(20.51) )

0
2
21


Notice: how it simpy returns the integer value of the float.

Alternatively, one can specify the number of decimal places using 'round(number[, ndigits])'.

In [2]:
number_list = [2.35, 1.25, 1.675, 10, 1.555E-9]

for number in number_list:
    number_rounded = round(number, 2)
    
    print('When rounded to 2 decimals, the number {0} becomes {1}. '.format(number, number_rounded))

When rounded to 2 decimals, the number 2.35 becomes 2.35. 
When rounded to 2 decimals, the number 1.25 becomes 1.25. 
When rounded to 2 decimals, the number 1.675 becomes 1.68. 
When rounded to 2 decimals, the number 10 becomes 10. 
When rounded to 2 decimals, the number 1.555e-09 becomes 0.0. 


Notice two things:

1. 10, which as no decimals initially and thus a questionable number of sig figs, stays as 10.
2. 1.555E-9 has **4 sig figs**, but the function round returned **0.0** -- which means we **lost** some data information!

Thus, Python3's built-in round function uses **Round Half To Even**<br><br>

**However**, notice the following weird result when we round 2.675 to two decimal places.<br><br>

Okay, one more thing to consider here, what would you expect from the following statement:

In [3]:
print( round(2.675, 2) )

2.67


If one considers the above **'round(1.675, 2)'** statement, the **'round(2.675, 2)'** should result in **2.68**. Instead we obtained **2.67**!<br><br>

Another and simpler example of this is:

In [4]:
0.1 + 0.1 + 0.1

0.30000000000000004

Reason: **floating-point representation error** (https://docs.python.org/3/tutorial/floatingpoint.html)

This is not a Python issue, but has to do with how computers store floating-point numbers in memory.

From a perspective of being exact and correct, this is something that one needs to be aware of -- particularly when significant figures are important -- when developing code for scientific usage.

---
### Python's math library (with examples)
(bias to **Round Up** and **Round Down**)

In [5]:
import math
## ceil stands for ceiling

print( math.ceil(1.2) )
print( math.ceil(2) )
print( math.ceil(-0.5) )

2
2
0


In [6]:
print( math.floor(1.2) )
print( math.floor(2) )
print( math.floor(-0.5) )

1
2
-1


---
### Python's decimal library

In [7]:
import decimal
print( decimal.Decimal(2.675) )
print( decimal.Decimal('2.675') )

2.67499999999999982236431605997495353221893310546875
2.675


In [8]:
decimal.Decimal(2.675) + decimal.Decimal(2.675)

Decimal('5.349999999999999644728632120')

In [9]:
decimal.Decimal('2.675') + decimal.Decimal('2.675')

Decimal('5.350')

In [10]:
## Rounding witih decimals library

## quantized - using Rounding Half to Even
print( decimal.Decimal("2.675").quantize(decimal.Decimal("1.00")) ) ## The 1.00 signifies the number of places

2.68


In [11]:
## How to automatically do it when performing math
decimal.getcontext().prec = 2

## 1.23 + 2.32 = 3.55
print( decimal.Decimal("1.23") + decimal.Decimal("2.32") )

3.6


The decimal library has lots of options:

decimal.ROUND_CEILING $\rightarrow$ Rounding up

decimal.ROUND_FLOOR$\rightarrow$ Rounding down

decimal.ROUND_DOWN$\rightarrow$ Truncation

decimal.ROUND_UP$\rightarrow$ Rounding away from zero

decimal.ROUND_HALF_UP$\rightarrow$ Rounding half away from zero

decimal.ROUND_HALF_DOWN$\rightarrow$ Rounding half towards zero

decimal.ROUND_HALF_EVEN$\rightarrow$ Rounding half to even

decimal.ROUND_05UP$\rightarrow$ Rounding towards zero, unless the last digit is 0 or 5 then round away from zero

In [12]:
## make the code more readable by changing your import statement
from decimal import Decimal

decimal.getcontext().rounding = decimal.ROUND_CEILING
print( Decimal('2.675').quantize(Decimal('1.0')) )
print( Decimal('-2.675').quantize(Decimal('1.0')) )

2.7
-2.6


In [13]:
decimal.getcontext().prec = 3  ## Do this otherwise you get errors from the first print statement
decimal.getcontext().rounding = decimal.ROUND_FLOOR
print( Decimal('2.675').quantize(Decimal('1.00')) )
print( Decimal('-2.675').quantize(Decimal('1.00')) )

2.67
-2.68


---
### Python's numpy library

In [14]:
import numpy

print( numpy.around(1.5) )
print( numpy.around(2.5) )
print( numpy.around(2.51) )

print()
print( numpy.around(2.35, decimals=1) )
print( numpy.around(1.25, decimals=1) )
print( numpy.around(2.665, decimals=2) )
print( numpy.around(2.675, decimals=2) )

2.0
2.0
3.0

2.4
1.2
2.66
2.68


Thus, Numpy's around function even better using **Round Half To Even**

**Note: Round Half To Even is usually the default (https://en.wikipedia.org/wiki/IEEE_754#Rounding_rules).**

---
Now, there is one more way to consider...

### Formating print statements containing floats

https://docs.python.org/3.1/library/string.html#grammar-token-precision

In [22]:
## g: General format
##    - fixed point or exponential
print('{0:g}, {1:g}, {2:g}'.format(3.4500045, 3.5500055, 300000000005))

3.45, 3.55001, 3e+11


In [27]:
## f: Fixed point
## rounding to odd number
##    demo: change 1f to 3f in on value
print('{0:.1f}, {1:.1f}, {2:.1f}, {3:.1f}, {4:.1f}'.format(3.45, 3.55, -3.45, -3.55, 300000000005))

3.5, 3.5, -3.5, -3.5, 300000000005.0


In [25]:
## e: Exponent notation
## rounding to odd number
print('{0:.1e}, {1:.1e}, {2:.1e}'.format(3.45, 3.55, 300000000005))

3.5e+00, 3.5e+00, 3.000e+11


In [31]:
print(f'Average value: {3.45:.1f}')

Average value: 3.5


In [29]:
print(f'Average value: {-3.55:.1f}')

Average value: -3.5


In [30]:
number_list = [2.35, 1.25, 1.675, 10, 1.555E-9]

for number in number_list:
    number_rounded = '{:0.2f}'.format(number)
    print('When rounded to 2 decimals, the number {0} becomes {1}. '.format(number, number_rounded))

When rounded to 2 decimals, the number 2.35 becomes 2.35. 
When rounded to 2 decimals, the number 1.25 becomes 1.25. 
When rounded to 2 decimals, the number 1.675 becomes 1.68. 
When rounded to 2 decimals, the number 10 becomes 10.00. 
When rounded to 2 decimals, the number 1.555e-09 becomes 0.00. 


 Notice one important thing from the above example:
 1. Strickly speaking, 10 has an undertermined number of significant figures, or if one assumes that a decimal should be present (i.e. 10.), then 2 significant figures. The fixed (i.e. f) formatting results in a number with 4 significant figures -- which means we  **artificially/erroneously gained** some data information!
 
 Let's check out what general (i.e. g) formatting does:

In [21]:
for number in number_list:
    number_rounded = '{:0.2g}'.format(number)
    print('When rounded to 2 decimals, the number {0} becomes {1}. '.format(number, number_rounded))

When rounded to 2 decimals, the number 2.35 becomes 2.4. 
When rounded to 2 decimals, the number 1.25 becomes 1.2. 
When rounded to 2 decimals, the number 1.675 becomes 1.7. 
When rounded to 2 decimals, the number 10 becomes 10. 
When rounded to 2 decimals, the number 1.555e-09 becomes 1.6e-09. 


## Take-home message

You have **think critically** about what you are doing and what your final goal is.

In other words, don't blindly take information that you find quickly and blindly use it.

**Create short test codes** to **verify** that **'what you think you are doing'** is actually **'what you are doing!'**